blocksync/buffer.go

```diff
@@ -74,25 +74,35 @@ func (b *blockBuffer) AddBlock(tipHeight uint64, blk *block.Block) {
 	if blkHeight > tipHeight && blkHeight <= tipHeight+b.bufferSize {
 		b.mu.Lock()
 		defer b.mu.Unlock()
-
-		if _, ok := b.blocks[blkHeight]; !ok {
-			b.blocks[blkHeight] = blk
+		if _, ok := b.blockQueues[blkHeight]; !ok {
+			b.blockQueues[blkHeight] = &uniQueue{
+				blocks: []*block.Block{},
+				hashes: map[hash.Hash256]bool{},
+			}
 		}
+		b.blockQueues[blkHeight].enque(blk)
```

Comment:
one thought from that wrong block (timestamp is 2 months back) you found:
we can save the last successfully committed block's time, and accept incoming blocks only if timestamp is later than that
```
if blk.Time.Before(b.tipTime) { continue }
```
b.tipTime is updated every time a block is successfully committed